Library/Homebrew/cmd/update.rb

```diff
@@ -277,8 +277,7 @@ def pull!(options = {})
     reset_on_interrupt { safe_system "git", *args }
 
     @current_revision = read_current_revision
-
-    if @initial_branch != @upstream_branch && !@initial_branch.empty?
+    if ARGV.homebrew_developer? && @initial_branch != @upstream_branch && !@initial_branch.empty?
```

Comment:
I'd definitely like to avoid popping the stash for end-users. Note that it only prints the message unless you were already on a non-default branch; I think we can just print the message in that case too.